static/js/message_list_data_cache.js

```diff
@@ -0,0 +1,84 @@
+import {Filter} from "./filter";
+import * as hash_util from "./hash_util";
+
+export class MessageListDataCache {
+    constructor() {
+        this._objects = new Map();
+    }
+
+    _get_key(filter) {
+        // The MLDCache API uses the filter object as the key.
+        // Internally, we serialize it, so this function
+        // should be called when only we want to access or
+        // create a stored MLD object.
+        return hash_util.operators_to_hash(filter.operators());
+    }
+
+    get(filter) {
+        const key = this._get_key(filter);
+        return this._objects.get(key);
+    }
+
+    add(mld) {
+        if (!mld.filter.can_apply_locally() || mld.empty()) {
+            return;
+        }
+        const key = this._get_key(mld.filter);
+        this._objects.set(key, mld);
+    }
+
+    _get_by_key(key) {
+        return this._objects.get(key);
+    }
+
+    _get_key_from_operators(raw_operators) {
+        const filter = new Filter(raw_operators);
+        const operators = filter.operators();
+        return hash_util.operators_to_hash(operators);
+    }
+
+    get_valid_mlds(filter) {
+        const operators = filter.operators();
```

Comment:
Can you add a comment explaining what policy this is trying to implement?  I wonder if we want something like "is_superset(filter1, filter2)" as the core logic bit of it.